### PR TITLE
Extend pipeline to test HA scenario

### PR DESCRIPTION
**Pull request by rukawata**
This PR is to address #365.

#### Checklist

- [x] Add new use case v2-allNew-HA with manual IP assignment for HANA servers.
- [x] Modify existing use case naming to reflect the fact they are for single node, eg. v2-allNew-SN.
- [x] Update documents/comments with the above changes.
- [x] Ensure pipeline runs without any problem for new PRs.

#### In addition:
- [x] Modified python script so we could merge list in JSON.
- [x] Adjust job name to use '_' for more readability. ('-' is not allowed in the job name) 
- [x] Prepare the pipeline agent to use 18.04
- [x] Ansible installation is no longer required, therefore remove that.

### CODE REVIEW

File: azure-pipelines-v2-daily.yaml [hunk 1]
```diff
@@ -27,29 +27,29 @@ variables:
...
-# - Below testcases are tested only with SLES12 SP4                 # 
-#   - createAllNew: it creates all resources from scratch           # 
-#   - reuseRG: it only reuses an existing resource group            #
-#   - reuseVnet: it only reuses an existing vnet from createAllNew  #
-#   - reuseNSG: it only reuses an existing NSG from createAllNew    #
-# - createAllNew testcase will be test again with below OS          #
-#   - RHEL7.3                                                       #
-#####################################################################
+###############################################################################
+# This stage create resources for below test cases:                           #
```
Comment (tniek) [inline azure-pipelines-v2-daily.yaml]: "creates"

File: azure-pipelines-v2-daily.yaml [hunk 2]
```diff
@@ -27,29 +27,29 @@ variables:
...
+# - Below testcases are tested only with SLES12 SP5                           #
+#   - createAllNew-HA: it creates all resources from scratch in HA Scenario   # 
+#   - createAllNew-SN: it creates all resources from scratch for single node  # 
+#   - reuseRG-SN: it only reuses an existing resource group for single node   #
+#   - reuseVnet-SN: it only reuses an existing vnet from createAllNew-SN      #
+#   - reuseNSG-SN: it only reuses an existing NSG from createAllNew-SN        #
+# - createAllNew-SN testcase will be test again with below OS                 #
```
Comment (tniek) [inline azure-pipelines-v2-daily.yaml]: "will be tested"

File: azure-pipelines-v2.yaml [hunk 3]
```diff
@@ -21,29 +21,29 @@ variables:
...
-# - Below testcases are tested only with SLES12 SP4                 # 
-#   - createAllNew: it creates all resources from scratch           # 
-#   - reuseRG: it only reuses an existing resource group            #
-#   - reuseVnet: it only reuses an existing vnet from createAllNew  #
-#   - reuseNSG: it only reuses an existing NSG from createAllNew    #
-# - createAllNew testcase will be test again with below OS          #
-#   - RHEL7.3                                                       #
-#####################################################################
+###############################################################################
+# This stage create resources for below test cases:                           #
```
Comment (tniek) [inline azure-pipelines-v2.yaml]: "creates"

File: azure-pipelines-v2.yaml [hunk 4]
```diff
@@ -21,29 +21,29 @@ variables:
...
+# - Below testcases are tested only with SLES12 SP5                           #
+#   - createAllNew-HA: it creates all resources from scratch in HA Scenario   # 
+#   - createAllNew-SN: it creates all resources from scratch for single node  # 
+#   - reuseRG-SN: it only reuses an existing resource group for single node   #
+#   - reuseVnet-SN: it only reuses an existing vnet from createAllNew-SN      #
+#   - reuseNSG-SN: it only reuses an existing NSG from createAllNew-SN        #
+# - createAllNew-SN testcase will be test again with below OS                 #
```
Comment (tniek) [inline azure-pipelines-v2.yaml]: "will be tested"

File: templates/job-template-allReuseCases.yaml [hunk 5]
```diff
@@ -3,10 +3,10 @@ parameters:
   osImage: ''
 
 jobs:
-- job: createAllNew${{parameters.osVersion}}
+- job: createAllNew_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: Proper branching (if/else) logic would make this easier to read.
What are the possible values for `parameters.osVersion`?

File: templates/job-template-allReuseCases.yaml [hunk 6]
```diff
@@ -19,10 +19,26 @@ jobs:
     parameters:
       testCaseName: $(testcase)
       osImage: ${{parameters.osImage}}
-- job: reuseRG${{parameters.osVersion}}
+- job: createAllNew_HA_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: See above; not a fan of concatenating the job name

File: templates/job-template-allReuseCases.yaml [hunk 7]
```diff
@@ -19,10 +19,26 @@ jobs:
...
   timeoutInMinutes: 180
   variables:
-    scenario: v2-reuseRG
+    scenario: v2-allNew-HA
+    testcase: $(scenario)-${{parameters.osVersion}}-$(Build.BuildId)
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: See above

File: templates/job-template-allReuseCases.yaml [hunk 8]
```diff
@@ -48,16 +64,16 @@ jobs:
   - template: ansible-deployment-steps.yaml
     parameters:
       testCaseName: $(testcase)
-- job: reuseVnet${{parameters.osVersion}}
-  dependsOn: createAllNew${{parameters.osVersion}}
+- job: reuseVnet_SN_${{parameters.osVersion}}
+  dependsOn: createAllNew_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: See above

File: templates/job-template-allReuseCases.yaml [hunk 9]
```diff
@@ -74,16 +90,16 @@ jobs:
     parameters:
       testCaseName: $(testcase)
       osImage: ${{parameters.osImage}}
-- job: reuseNSG${{parameters.osVersion}}
+- job: reuseNSG_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: See above

File: templates/job-template-allReuseCases.yaml [hunk 10]
```diff
@@ -74,16 +90,16 @@ jobs:
...
   timeoutInMinutes: 180
-  dependsOn: createAllNew${{parameters.osVersion}}
+  dependsOn: createAllNew_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-allReuseCases.yaml]: See above

File: templates/job-template-per-os-daily.yaml [hunk 11]
```diff
@@ -3,10 +3,10 @@ parameters:
   osImage: ''
 
 jobs:
-- job: createAllNew${{parameters.osVersion}}
+- job: createAllNew_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-per-os-daily.yaml]: See above

File: templates/job-template-per-os.yaml [hunk 12]
```diff
@@ -3,10 +3,10 @@ parameters:
   osImage: ''
 
 jobs:
-- job: createAllNew${{parameters.osVersion}}
+- job: createAllNew_SN_${{parameters.osVersion}}
```
Comment (tniek) [inline templates/job-template-per-os.yaml]: See above

File: templates/tempGen/tempGen.py [hunk 13]
```diff
@@ -7,7 +7,11 @@ def mergeJSON(a, b, path=None):
     for key in b:
         if key in a:
             if isinstance(a[key], dict) and isinstance(b[key], dict):
-                mergeJSON(a[key], b[key], path + [str(key)])
+                t = mergeJSON(a[key], b[key], path + [str(key)])
```
Comment (tniek) [inline templates/tempGen/tempGen.py]: Generic variable names (a, b, t) makes it hard to read.
Can you replace them with more descriptive variable names?

File: templates/job-template-allReuseCases.yaml (hunk 5 above)
Comment (rukawata) [reply templates/job-template-allReuseCases.yaml]: What if/else logic you are referring here?

As for parameters.osVersion, please check [here](https://github.com/Azure/sap-hana/blob/3e5634bb432598a07b5d4e93ca96fde59d917f18/azure-pipelines-v2.yaml#L50-L61)

File: templates/tempGen/tempGen.py (hunk 13 above)
Comment (rukawata) [reply templates/tempGen/tempGen.py]: Changed.

**Pull request merged**